Core/ConnectedShapeBasedInterpolation.h

```diff
@@ -0,0 +1,220 @@
+		connectivity->Update();
+		num_objects = connectivity->GetObjectCount();
+		return connectivity->GetOutput();
+	}
+
+	/// compute center of mass of foreground object, assumes BG=0
+	std::vector<itk::Point<double, 2>> get_center_of_mass(const mask_type* img, int num_objects) const
+	{
+		using idx_type = itk::ContinuousIndex<double, 2>;
+		std::vector<double> n(num_objects, 0.0);
+		std::vector<idx_type> idx(num_objects, idx_type({0,0}));
+
+		itk::ImageRegionConstIteratorWithIndex<mask_type> it(img, img->GetBufferedRegion());
+		for (it.GoToBegin(); !it.IsAtEnd(); ++it)
+		{
+			int const id = static_cast<int>(it.Get()) - 1;
+			if (id >= 0)
+			{
+				n[id]++;
+				idx[id][0] += it.GetIndex()[0];
+				idx[id][1] += it.GetIndex()[1];
+			}
+		}
+
+		std::vector<itk::Point<double, 2>> pts(num_objects);
+		for (int id=0; id<num_objects; ++id)
+		{
+			if (n[id] > 0)
+			{
+				idx[id][0] /= n[id];
+				idx[id][1] /= n[id];
+			}
+			img->TransformContinuousIndexToPhysicalPoint(idx[id], pts[id]);
+		}
+		return pts;
+	}
+
+	/// shift image by a specified amount
+	image_type::Pointer move_image(const image_type* img, const itk::Vector<double,2>& translation) const
+	{
+		using resample_filter_type = itk::ResampleImageFilter<image_type, image_type>;
+		using translation_type = itk::TranslationTransform<double, 2>;
+		
+		auto resample_filter = resample_filter_type::New();
+		resample_filter->SetInput(img);
+		
+		auto nn_interpolator = itk::NearestNeighborInterpolateImageFunction<image_type>::New();
+		resample_filter->SetInterpolator(nn_interpolator);
+
+		auto transform = translation_type::New();
+		transform->Translate(translation);
+		resample_filter->SetTransform(transform.GetPointer());
+
+		resample_filter->SetOutputParametersFromImage(img);
+		resample_filter->Update();
+		return resample_filter->GetOutput();
+	}
+
+	/// compute signed distance function, assumes BG=0
+	image_type::Pointer compute_sdf(const image_type* img) const
+	{
+		using distance_filter_type = itk::SignedMaurerDistanceMapImageFilter<image_type, image_type>;
+		auto dt_filter1 = distance_filter_type::New();
+		dt_filter1->SetInput(img);
+		dt_filter1->SetBackgroundValue(0); // ?
+		dt_filter1->SetInsideIsPositive(false);
+		dt_filter1->SetSquaredDistance(false);
+		dt_filter1->SetUseImageSpacing(true);
+		dt_filter1->Update();
+		return dt_filter1->GetOutput();
+	}
+
+	/// interpolate between known slices
+	image_type::Pointer interpolate(image_stack_type* image3d, double ratio_12, const itk::Vector<double, 2>& translation_12) const
+	{
+		auto id = itk::IdentityTransform<double, 3>::New();
+		auto origin = image3d->GetOrigin();
+		origin[2] = ratio_12;
+		auto size = image3d->GetBufferedRegion().GetSize();
+		size[2] = 1;
+
+		using image_3d_type = itk::Image<float, 3>;
+		auto resample_filter = itk::ResampleImageFilter<image_stack_type, image_3d_type>::New();
+		resample_filter->SetTransform(id.GetPointer());
+		resample_filter->SetInput(image3d);
```

Comment:
I think all these filters also have a set multi threaded function (do not remember the exact syntax which can be of intereset performance wise)

Reply:
I will double-check, however, I thought they follow global options, e.g. based on environment variables, etc.

Reply:
that may also be.